packages/sequelize/src/index.ts

```diff
@@ -0,0 +1,156 @@
+import type { Account as ApadterAccount } from "next-auth"
+import type {
+  Adapter,
+  AdapterUser,
+  AdapterSession,
+  VerificationToken,
+} from "next-auth/adapters"
+import { Sequelize, Model, ModelCtor } from "sequelize"
+import {
+  accountSchema,
+  userSchema,
+  sessionSchema,
+  verificationTokenSchema,
+} from "./schema"
+
+// @see https://sequelize.org/master/manual/typescript.html
+interface AccountInstance
+  extends Model<ApadterAccount, Partial<ApadterAccount>>,
+    ApadterAccount {}
+interface UserInstance
+  extends Model<AdapterUser, Partial<AdapterUser>>,
+    AdapterUser {}
+interface SessionInstance
+  extends Model<AdapterSession, Partial<AdapterSession>>,
+    AdapterSession {}
+interface VerificationTokenInstance
+  extends Model<VerificationToken, Partial<VerificationToken>>,
+    VerificationToken {}
+
+export { accountSchema, userSchema, sessionSchema, verificationTokenSchema }
+
+export default function SequelizeAdapter(client: Sequelize): Adapter {
+  const tableOptions = { underscored: true, timestamps: false }
+
+  const Account =
+    (client.models.Account as ModelCtor<any>) ||
+    client.define<AccountInstance>("account", accountSchema, tableOptions)
+  const User =
+    (client.models.User as ModelCtor<any>) ||
+    client.define<UserInstance>("user", userSchema, tableOptions)
+  const Session =
+    (client.models.Session as ModelCtor<any>) ||
+    client.define<SessionInstance>("session", sessionSchema, tableOptions)
+  const VerificationToken =
+    (client.models.VerificationToken as ModelCtor<any>) ||
+    client.define<VerificationTokenInstance>(
+      "verificationToken",
+      verificationTokenSchema,
+      tableOptions
+    )
+
+  Account.belongsTo(User, { onDelete: "cascade" })
+  Session.belongsTo(User, { onDelete: "cascade" })
```

Comment:
Is this necessarily done runtime? I am not familiar with Sequelize, but usually, you would do some kind of seeding/setup step to define the database schema, right? Does Sequelize have anything similar to TypeORM's `syncrhonize: true` (https://github.com/typeorm/typeorm/blob/master/docs/faq.md#how-do-i-update-a-database-schema)?

We have to be careful not to update things in production to prevent data loss. In development though, auto-syncing model changes might be useful. #224 how we handled this for the TypeORM adapter recently.

Reply:
Sequelize has the [`sequelize.sync()`](https://sequelize.org/master/manual/model-basics.html#model-synchronization) function similar to TypeORM's synchronize option. Defining these associations at runtime is recommended and doesn't produce any SQL unless `sync()` is called.

It would be best practice for the user to define the tables as migrations, but sequelize will automatically create the necessary tables/FKs/indexes with `sync` like so:


```
const sequelize = new Sequelize('sqlite:memory:')
const adapter = SequelizeAdapter(sequelize)

sequelize.sync()

export default NextAuth({
  ...
  adapter
  ...
})
```

I've added a section in the README about how to create tables like this with a mention that best practice is still migrations in e0b1dae.

Reply:
I think we should default to synchronizing if we detect `process.env.NODE_ENV !== "production"`, but make it so that users can opt-out even then, with a config option like `synchronize: false`.

We could reuse this warning by some rewording: https://next-auth.js.org/warnings#adapter_typeorm_updating_entities

I would add `client.sync()` in an if statement with the warning	similar to this [`c2559b7` (#224)](https://github.com/nextauthjs/adapters/pull/224/commits/c2559b7557e7786db8dc96d7f545a9c80c79af18) in the body of the Adapter.

Otherwise, we should create explain how to migrate manually: https://sequelize.org/master/manual/migrations.html

Reply:
@balazsorban44 I'm thinking about how we might achieve this. I remembered that `sync` actually returns a promise, in which case we'd have to make the adapter `async` which would mess with the user-facing API.

eg.
```ts
if (process.env.NODE_ENV !== 'production' && synchronize) {
  await Promise.all([
    User.sync(),
    Account.sync(),
    Session.sync(),
    VerificationToken.sync()
  ])
}
```

I'm wondering if including `sync`, which is potentially dangerous or unexpected from the user's perspective, is worth it in this regard? There's not a huge advantage over just leaving this to user land code, right?

Reply:
I see. Ultimately, this is why we had to get a connection manager in each adapter method with TypeORM, because getting a manager was an async operation. :sweat_smile: 

https://github.com/nextauthjs/adapters/blob/next/packages/typeorm-legacy/src/index.ts

I don't want the adapter to be async as that would complicate the initialization.

I see there is an all-in-one `.sync()` method as well.

https://sequelize.org/master/class/lib/sequelize.js~Sequelize.html#instance-method-sync

My idea is that we should create a wrapper for it, and add a `sync: boolean | SyncOptions` to the adapter options, and do something similar to TypeORM, and call it in each method first.

I feel like we should stretch here, as the DX would be better. Especially useful if you would need to debug, I think.

I am thinking of something like this:
```ts
function SequelizeAdapter(client, options) {
  // Make sure we only sync once per invocation
  let _synced = false
  const sync = async () => {
    if (
      process.env.NODE_ENV === "production" ||
      _synced ||
      options.sync === false
    )
      return

    const syncOptions =
      typeof options.sync === "object"
        ? options.sync
        : {
            // ... some default options, if we need it
          }
    await client.sync(syncOptions)
    _synced = true
  }

  return {
    async createUser() {
      await sync()
      //...
    },
  }
}
```

Reply:
Yeah I think if we want to do this this is probably the best option. I've done this in 48961d1. Unfortunately it's not the prettiest solution.

If this is occurring in multiple adapters I think it's reasonable to consider an optional `connect` or `init` function returned from an adapter which is then called by the main next-auth package prior to calling the other functions.

Reply:
Most adapters seem to connect/disconnect automatically, actually. In my experience, it is only TypeORM that needs manual connection handling so far. :confused: